[r2r] Add nft docs

smk762 opened the pull request:
Closes https://github.com/KomodoPlatform/komodo-docs-mdx/issues/43

- [x] get_nft_list
- [x] get_nft_transfers
- [x] get_nft_metadata
- [x] update_nft
- [x] withdraw_nft (erc721)
- [x] withdraw_nft (erc1155)
- [x] clearing the nft database tables
- [x] querying the nft database tables

laruh commented on src/pages/atomicdex/api/v20-dev/non_fungible_tokens/index.mdx:
```diff
@@ -0,0 +1,577 @@
+
+| Parameter     | Type             | Description                                                                                               |
+| ------------- | ---------------- | ----------------------------------------------------------------------------------------------------------|
+| chains        | array            | List of chains to scan for NFTs. Chains must be [activated](/atomicdex/api/legacy/coin_activation) first. |
```
User doesnt have to activate platform coin to see their NFT list, tx history or nft metadata.
Right now it is required only for withdraw nft  methods.

smk762 replied:
removed, thanks.

laruh replied:
please remove `Chains must be activated` from [get tx history](https://github.com/KomodoPlatform/komodo-docs-mdx/pull/42/files#diff-952348f4452c003b588224faac88b718b4b5a1d7c957b43a8ec35c44178d44bdR168) too

and add info that user needs to activate platform coin in [nft withdraw section](https://github.com/KomodoPlatform/komodo-docs-mdx/pull/42/files#diff-952348f4452c003b588224faac88b718b4b5a1d7c957b43a8ec35c44178d44bdR373)

smk762 replied:
Removed from all tables; added a note for withdraw method.

laruh commented on src/pages/atomicdex/api/v20-dev/non_fungible_tokens/index.mdx:
```diff
@@ -0,0 +1,577 @@
+| Parameter     | Type             | Description                                                                                               |
+| ------------- | ---------------- | ----------------------------------------------------------------------------------------------------------|
+| chains        | array            | List of chains to scan for NFTs. Chains must be [activated](/atomicdex/api/legacy/coin_activation) first. |
+| url           | string           | URL link to [Moralis API](https://docs.moralis.io/) (or proxy).                                           |
```
Moralis nft endpoint `https://deep-index.moralis.io/` wont work, bcz we removed moralis API-KEY owned by user from mm config.

Also, if user check https://docs.moralis.io link they wont understand what they need to find there and which moralis endpoint they should provide. In url we use proxy url `https://moralis-proxy.komodo.earth` instead of this part of full url for moralis requests.
For the future its better to provide link to moralis reference page directly https://docs.moralis.io/web3-data-api/evm/reference#get-nfts

![image](https://github.com/KomodoPlatform/komodo-docs-mdx/assets/52405288/d3d78bca-3ac0-4042-928f-48439561a0b8)

smk762 replied:
Reworded, and added note about future potential param for user API key.

laruh reviewed (changes requested):
Great work!
This is first review iteration, I will provide more notes later.

laruh commented on src/pages/atomicdex/api/v20-dev/non_fungible_tokens/index.mdx:
```diff
@@ -0,0 +1,577 @@
+
+### Response Parameters
+
+| Parameter            | Type             | Description                                                                                      |
+| ---------------------| ---------------- | ------------------------------------------------------------------------------------------------ |
+| amount               | string           | Amount of tokens.                                                                                |
+| block_number_minted  | integer          | Block height when the token was minted.                                                          |
+| block_number         | integer          | Current block height.                                                                            |
+| chain                | string           | Chain name. One of `AVALANCHE`, `BSC`, `ETH`, `FANTOM`, or `POLYGON`.                            |
+| collection_name      | string           | Collection name. May be `null`.                                                                  |
+| contract_type        | string           | Contract type. `ERC721` or `ERC1155`.                                                            |
+| owner_of             | string           | Address of the token owner.                                                                      |
+| token_address        | string           | Token address.                                                                                   |
+| token_id             | string           | Token ID.                                                                                        |
+| token_hash           | string           | Token hash.                                                                                      |
+| symbol               | string           | Token symbol.                                                                                    |
+| token_uri            | string           | Token URI.                                                                                       |
+| metadata             | string           | Token metadata.                                                                                  |
+| last_token_uri_sync  | string           | Last time the token URI was synced.                                                              |
+| last_metadata_sync   | string           | Last time the token metadata was synced.                                                         |
+| minter_address       | string           | Minter address.                                                                                  |
+| possible_spam        | boolean          | Possible spam.                                                                                   |
+| uri_meta             | object           | A Standard [TokenMetadata](LINK TO COMMON STRUCTURE) object.                                     |
```
This is how Nft structure currently looks like in [add-token-uri-info-to-nft-struct ](https://github.com/KomodoPlatform/atomicDEX-API/tree/add-token-uri-info-to-nft-struct) branch. This is not in dev yet, but GUI team uses this branch bcz they need `UriMeta`
you can check which fields can be null
```rust
pub struct Nft {
    pub(crate) chain: Chain,
    pub(crate) token_address: String,
    pub(crate) token_id: BigDecimal,
    pub(crate) amount: BigDecimal,
    pub(crate) owner_of: String,
    pub(crate) token_hash: String,
    pub(crate) block_number_minted: u64,
    pub(crate) block_number: u64,
    pub(crate) contract_type: Option<ContractType>,
    pub(crate) collection_name: Option<String>,
    pub(crate) symbol: Option<String>,
    pub(crate) token_uri: Option<String>,
    pub(crate) metadata: Option<String>,
    pub(crate) last_token_uri_sync: Option<String>,
    pub(crate) last_metadata_sync: Option<String>,
    pub(crate) minter_address: Option<String>,
    pub(crate) possible_spam: Option<bool>,
    pub(crate) uri_meta: UriMeta,
}
```
```rust
pub(crate) struct UriMeta {
    image: Option<String>,
    #[serde(rename(deserialize = "name"))]
    token_name: Option<String>,
    description: Option<String>,
    attributes: Option<Json>,
    animation_url: Option<String>,
}
```
ContractType is enum and yeah it could be ERC721 or ERC1155 (uppercase)

smk762 replied:
Updated, using descriptions from https://docs.moralis.io/web3-data-api/evm/reference/get-wallet-nfts?address=0xd8da6bf26964af9d7eed9e03e53415d37aa96045&chain=eth&format=decimal&token_addresses=[]&media_items=false

laruh commented on src/pages/atomicdex/api/v20-dev/non_fungible_tokens/index.mdx:
```diff
@@ -0,0 +1,577 @@
+| contract_type        | string           | Contract type. `ERC721` or `ERC1155`.                                                            |
+| token_address        | string           | Address of token transferred.                                                                    |
+| token_id             | string           | Token ID.                                                                                        |
+| possible_spam        | boolean          | Possible spam.  MORE  INFO?                                                                      |
```
You can provide link to [moralis feature description ](https://docs.moralis.io/web3-data-api/evm/nft-spam-detection). So there is the info

smk762 replied:
Done.

laruh commented on src/pages/atomicdex/api/v20-dev/non_fungible_tokens/index.mdx:
```diff
@@ -0,0 +1,577 @@
+| transaction_type     | string           | Transation type. `Single` or `WHAT IS THE OTHER TYPE?`.                                          |
+| from_address         | string           | Address of previous owner which sent the token(s).                                               |
+| to_address           | string           | Address of new owner which received the token(s).                                                |
+| status               | string           | Transfer status. WHAT ARE THE POSSIBLE VALUES?                                                   |
```
Receive or Send

```rust
pub(crate) enum TransferStatus {
    Receive,
    Send,
}
```

It basically shows that if from_address was the my_address (weans our user address), then user sent nft,
If to_address == user address, then user received nft in transaction.
It was made for GUI team and also it enable user not to check from or to address with their wallet address to understand in this transaction they received or sent nft.

```rust
let status = if wallet_address.to_lowercase() == transfer_wrapper.to_address {
                        TransferStatus::Receive
                    } else {
                        TransferStatus::Send
                    };
```

smk762 replied:
thanks, updated description

laruh commented on src/pages/atomicdex/api/v20-dev/non_fungible_tokens/index.mdx:
```diff
@@ -0,0 +1,577 @@
+| coin                 | string           | Coin name. One of `AVALANCHE`, `BSC`, `ETH`, `FANTOM`, or `POLYGON`. CONFIRM: EVERYWHERE ELSE THIS PARAM IS `chain`.                            |
+| block_height         | integer          | Block height of the transaction. If the value is `0`, the transaction is unconfirmed.            |
+| timestamp            | integer          | Timestamp of the block containing the withdrawl transaction.                                     |
+| internal_id          | integer          | I HAVE NO IDEA WHAT THIS MEANS.                                                                  |
```
yeah, all our structures for Transaction details have this field.
`Internal MM2 id used for internal transaction identification, for some coins it might be equal to transaction hash`

This is how it looks in [TransactionNftDetails](https://github.com/KomodoPlatform/atomicDEX-API/blob/dev/mm2src/coins/nft/nft_structs.rs#L214)

When user generates nft transaction [I set 0](https://github.com/KomodoPlatform/atomicDEX-API/blob/dev/mm2src/coins/eth.rs#L980)

It is also empty when [user generates tx to withdraw ETH coins, this is for example, other coins also use this structure](https://github.com/KomodoPlatform/atomicDEX-API/blob/dev/mm2src/coins/eth.rs#L868). Actually this is the response for [withdraw method](https://github.com/KomodoPlatform/atomicDEX-API/blob/dev/mm2src/coins/eth.rs#L868)

But we fill it during [processing erc20 history](https://github.com/KomodoPlatform/atomicDEX-API/blob/dev/mm2src/coins/eth.rs#L2940)
here the [process_erc20_history function description ](https://github.com/KomodoPlatform/atomicDEX-API/blob/dev/mm2src/coins/eth.rs#L2608)

So the response for our general withdraw method also contains this field
https://docs.komodoplatform.com/basic-docs/atomicdex-api-20/withdraw.html but its just not documented.

And yeah, we fill it for our internal identification of transaction, zcoin also has it in Tx details and we use it for [history](https://docs.komodoplatform.com/basic-docs/atomicdex-api-20-dev/zhtlc_coins.html#z-coin-tx-history)

Right now it is always zero and actually I dont save TransactionNftDetails in db as we use info from moralis (NftTransferHistory), but may be in the future. I will need it...

smk762 replied:
Updated description

smk762 replied:
added devnote in case of future changes

laruh commented on src/pages/atomicdex/api/v20-dev/non_fungible_tokens/index.mdx:
```diff
@@ -0,0 +1,594 @@
+
+#### 📌 ERC1155 Withdraw Example
+
+If you are sending 2 or more NFTs, you must use the `withdraw_erc1155` withdraw type.
```
this sentence has wrong meaning. withdraw_types `erc1155` or `erc721` dont depend on how many tokens you want to withdraw. 
it depends on `contract type/contract standard/token standard`. NFTs which we support have a solid `contract_type`: `erc721` or `erc1155`.
So what I want to say, user must use withdraw_erc721 if they want to withdraw nft which has erc721 contract_type.
If nft has erc1155 standard, then user must use withdraw_erc1155.

If nft has "contract_type": "ERC721" it means that it is absolutely unique . It has only 1 owner and owner can own only 1 NFT of some token_id in some token_address (token address actually is a contract address).

If nft has "contract_type": "ERC1155", it means that this standard allows more than 1 users to own more than 1 nft with identical token_id. 

So withdraw_type in request form have different fields for two different token standards.

```rust
// this struct is actually what you build for nft withdraw request
#[derive(Clone, Deserialize)]
pub struct WithdrawNftReq {
    pub(crate) url: Url,
    pub(crate) withdraw_type: WithdrawNftType,
}

#[derive(Clone, Deserialize)]
#[serde(tag = "type", content = "withdraw_data")]
#[serde(rename_all = "snake_case")]
pub enum WithdrawNftType {
    WithdrawErc1155(WithdrawErc1155),
    WithdrawErc721(WithdrawErc721),
}

#[derive(Clone, Deserialize)]
pub struct WithdrawErc1155 {
    pub(crate) chain: Chain,
    pub(crate) to: String,
    pub(crate) token_address: String,
    pub(crate) token_id: BigDecimal,
    pub(crate) amount: Option<BigDecimal>,
    #[serde(default)]
    pub(crate) max: bool,
    pub(crate) fee: Option<WithdrawFee>,
}

#[derive(Clone, Deserialize)]
pub struct WithdrawErc721 {
    pub(crate) chain: Chain,
    pub(crate) to: String,
    pub(crate) token_address: String,
    pub(crate) token_id: BigDecimal,
    pub(crate) fee: Option<WithdrawFee>,
}
```

As you can see in `withdraw_type` field we dont have one struct, we have enum type which contains two enumerations, [not WithdrawData object](https://github.com/KomodoPlatform/komodo-docs-mdx/pull/42/files#diff-965f9228b14137d6315455493012e3ebadd2db27aa3428da3e7cec792ef8270aR330-R332). 

So for instance "type": "withdraw_erc721" means that it is enumeration `WithdrawErc721` from `enum WithdrawNftType` and `withdraw_data` contains field for WithdrawErc721 
```rust
pub struct WithdrawErc721 {
    pub(crate) chain: Chain,
    pub(crate) to: String,
    pub(crate) token_address: String,
    pub(crate) token_id: BigDecimal,
    pub(crate) fee: Option<WithdrawFee>,
}
```
we just dont specify name of `WithdrawErc721` in json, bcz back end will build itself the object from fields .
```json
"withdraw_type": {
      "type": "withdraw_erc721",
      "withdraw_data": {
        "chain": "BSC",
        "to": "0x6FAD0eC6bb76914b2a2a800686acc22970645820",
        "token_address": "0xfd913a305d70a60aac4faac70c739563738e1f81",
        "token_id": "214300044414"
      }
    }
```

So [here](https://github.com/KomodoPlatform/komodo-docs-mdx/pull/42/files#diff-965f9228b14137d6315455493012e3ebadd2db27aa3428da3e7cec792ef8270aR330-R343) you need to provide two variants of fields.
If contact type erc721 then: `chain, to, token_address, token_id, fee.`
If contact type erc1155 then:  `chain, to, token_address, token_id, amount, max, fee`

smk762 replied:
thanks for clarification, I see the last description was a bit ambiguous.

smk762 replied:
I've created two standard objects to reference this

laruh commented on src/pages/atomicdex/api/v20-dev/non_fungible_tokens/index.mdx:
```diff
@@ -0,0 +1,594 @@
+
+
+## Withdraw NFTs {{label : 'withdraw_nft', tag : 'API-v2'}}
+
```
I think from this line you should to specify two types of withdraw nft requests.
for erc721 tokens and for erc1155 tokens.

laruh commented on src/pages/atomicdex/api/v20/index.mdx:
````diff
@@ -251,4 +251,94 @@ The `TokensRequest` object includes the following items for a given coin or toke
+}
+```
+
+### WithdrawType
+
+The `WithdrawType` object is used for NFT withdrawals. It includes the following items for a given coin or token:
````
well, Im not sure that you need to mention `WithdrawType` name, as we actually dont have it in backend. We have `enum WithdrawNftType`.
well we have field  `withdraw_type: WithdrawNftType` for withdraw nft request and it refers only to NFT section.

I even dont think that you should to mention `withdraw_type: WithdrawNftType` in v20/index.mdx section
![image](https://github.com/KomodoPlatform/komodo-docs-mdx/assets/52405288/7481dcd8-7f9e-4ff6-88eb-9fac23db22ed), bcz it refers only to `nft_fungible_tokens`.

smk762 replied:
I've renamed this object to `WithdrawNftType`. The objects in docs exist just as a reference for any nested json objects, generally they are named arbitrarily.

laruh reviewed (changes requested):
Thank you, this is second iteration with another notes. I will check more changes from the first later (need more time to check fields types).

laruh commented on src/pages/atomicdex/api/v20-dev/non_fungible_tokens/index.mdx:
```diff
@@ -0,0 +1,597 @@
+| transaction_type     | string           | Transation type. Possible values are `Single`.                                          |
+| from_address         | string           | Address of previous owner which sent the token(s).                                               |
+| to_address           | string           | Address of new owner which received the token(s).                                                |
+| status               | string           | Transfer status. Will be either `Send` or `Receive`.                                             |
```
Please add a note, that if there is the situation that if user sent NFT to them self (from to addresses are  identical), then status also will be `Receive`
it is my comment from the code.
`if my_wallet == from_address && my_wallet == to_address it is incoming tx`

smk762 replied:
thanks, added.

laruh reviewed (changes requested):
one more note

smk762 commented:
I've moved any other questions / comments into between `DevComment` tags in the doc. This is not displayed to users, but acts as a reminder for items which may change or nee more info etc.

laruh commented on src/pages/atomicdex/api/v20-dev/non_fungible_tokens/index.mdx:
```diff
@@ -0,0 +1,610 @@
+
+
+<DevComment>
+For uri_meta, I think only the values under the `attributes` key are variable. The other keys are always present, but may be `null`. This should link to a standard object.
```
well as we use Json object as the type for attributes it can have different view.
```rust
pub(crate) struct UriMeta {
    pub(crate) image: Option<String>,
    #[serde(rename(deserialize = "name"))]
    pub(crate) token_name: Option<String>,
    description: Option<String>,
    attributes: Option<Json>,
    animation_url: Option<String>,
}
```
Actually fields for uri_meta are collected from link from token_uri parameter.

Form [this example](https://unft-metadata.rgc-external-stage.unioverse.com/getMetadataForCollectionAndTokenV3/Unioverse%20Collectibles/1265789) attributes is just one object, but [here](https://tikimetadata.s3.amazonaws.com/tiki_box.json) it is the array. So we decided just to try to parse the whole Json for attributes. bcz nft authors can add there any data.

smk762 replied:
Update to include a standard TokenMetadata object

laruh commented on src/pages/atomicdex/api/v20-dev/non_fungible_tokens/index.mdx:
```diff
@@ -0,0 +1,610 @@
+| operator             | string           | May be `null`.                                                                                   |
+
+<DevComment>
+`block timestamp` is a date time string, though I would expect a unix timestamp
```
Moralis provides it as ISO formatted string "block_timestamp": "2023-05-28T16:51:16.000Z", so I decided to live it as it is.

smk762 replied:
k, comment removed

laruh commented on src/pages/atomicdex/api/v20-dev/non_fungible_tokens/index.mdx:
```diff
@@ -0,0 +1,610 @@
+
+<DevComment>
+`block timestamp` is a date time string, though I would expect a unix timestamp
+`verified` has no description.
+What are the other possible values for `transaction_type`?
+What is `operator`?
+What does `verified` mean?
```
the moralis support team doesn't seem to know. hah. I will try to answer by myself.
![image](https://github.com/KomodoPlatform/komodo-docs-mdx/assets/52405288/667cc1f7-c428-4db3-8a13-d93a1712d489)

laruh replied:
![image](https://github.com/KomodoPlatform/komodo-docs-mdx/assets/52405288/cd56c462-9bba-45ce-b5ed-172426ee2b2a)
seems like they didnt remove this field but renamed to `verified_collection`?
```json
    {
      "block_number": "47961267",
      "block_timestamp": "2023-09-25T09:04:16.000Z",
      "block_hash": "0x0b256b4de5e82a8635e8a1bc3149d62c8d1332b6ad3b473a5ffab55a2071c301",
      "transaction_hash": "0x22dd86996e164c015b761e9b58b573ab37bf553ebd3ff73fd68399a117088321",
      "transaction_index": 75,
      "log_index": 471,
      "value": "0",
      "contract_type": "ERC1155",
      "transaction_type": "Single",
      "token_address": "0xc06a6ab4c2b0ed14ed5692adf324a9a463259c43",
      "token_id": "0",
      "from_address": "0x27ad1f808c1ef82626277ae38998afa539565660",
      "from_address_label": null,
      "to_address": "0xf622a6c52c94b500542e2ae6bcad24c53bc5b6a2",
      "to_address_label": null,
      "amount": "1",
      "verified": 0,
      "operator": "0x27ad1f808c1ef82626277ae38998afa539565660",
      "possible_spam": true,
      "verified_collection": false
    },
    {
      "block_number": "44506622",
      "block_timestamp": "2023-06-30T06:48:00.000Z",
      "block_hash": "0xcc40b0c8814a132b335f6f2c7cad25d85870ce671173057ee61a522df7b3fd79",
      "transaction_hash": "0x87dae6d10a60ecbe13d4a851e3589b2e320de57478f324564ab895de0731b43c",
      "transaction_index": 49,
      "log_index": 257,
      "value": "0",
      "contract_type": "ERC1155",
      "transaction_type": "Single",
      "token_address": "0x48c75fbf0452fa8ff2928ddf46b0fe7629cca2ff",
      "token_id": "5",
      "from_address": "0xf622a6c52c94b500542e2ae6bcad24c53bc5b6a2",
      "from_address_label": null,
      "to_address": "0xd37abcb646f21ac9abf9a01ffdee48704f18ad11",
      "to_address_label": null,
      "amount": "1",
      "verified": 1,
      "operator": "0xf622a6c52c94b500542e2ae6bcad24c53bc5b6a2",
      "possible_spam": false,
      "verified_collection": true
    }
```    

I think we can just include description from reference page
https://docs.moralis.io/web3-data-api/evm/reference/get-wallet-nft-transfers

they added response fields info
![image](https://github.com/KomodoPlatform/komodo-docs-mdx/assets/52405288/f649d91b-7c62-4afd-9097-0af14d13f5f2)
![image](https://github.com/KomodoPlatform/komodo-docs-mdx/assets/52405288/880cd359-92f9-42a1-893f-f41958b6d65a)

laruh commented on src/pages/atomicdex/api/v20-dev/non_fungible_tokens/index.mdx:
```diff
@@ -0,0 +1,610 @@
+
+<DevComment>
+May change in future. See https://github.com/KomodoPlatform/komodo-docs-mdx/pull/42#discussion_r1197745312
+Right now `internal_id` is always zero and actually I dont save TransactionNftDetails in db as we use info from moralis (NftTransferHistory), but may be in the future. I will need it...
```
sorry will this text be visible on the website?:D
I would like to have an anonymous text like this:

Right now `internal_id` is always zero and actually we dont save `TransactionNftDetails` in db as we use info from moralis (`NftTransferHistory` structure). However, perhaps in the future, in the process of refactoring, we might stop using the moralis API and start using the API of decentralized platforms. In this case this field may be needed.

smk762 replied:
DevComment tags are not shown to the public, just a way for us to add comments to docs for later.

laruh commented on src/pages/atomicdex/api/v20-dev/non_fungible_tokens/index.mdx:
```diff
@@ -0,0 +1,610 @@
+
+
+<DevComment>
+Is `transaction_type` always `NftTransfer`?
```
Yes it is.
transaction_type: TransactionType::NftTransfer,

laruh reviewed (commented):
Thank you, next iteration.
I provided some notes and answers.

Sorry, as [new pr with db support ](https://github.com/KomodoPlatform/atomicDEX-API/pull/1833)changes some requests, it also requires some changes in docs.
[gist](https://gist.github.com/laruh/248d652d88891f3cadb09fea3a441466#file-add-nft-cache-support-md)

smk762 commented on src/pages/atomicdex/api/v20/index.mdx:
```diff
@@ -251,4 +251,113 @@ The `TokensRequest` object includes the following items for a given coin or toke
+| to            | string           | Destination address to withdraw the token to.                                                             |
+| token_address | string           | Token address.                                                                                            |
+| token_id      | string           | Token ID.                                                                                                 |
+
```
added, thanks

smk762 commented on src/pages/atomicdex/api/v20-dev/non_fungible_tokens/index.mdx:
```diff
@@ -0,0 +1,610 @@
+| transaction_index    | string           | Transaction index of the transfer.                                                               |
+| log_index            | string           | Log index of the transfer.                                                                       |
+| value                | string           | Tranfer value.                                                                                   |
+| transaction_type     | string           | Transation type. Possible values are `Single`.                                                   |
```
In the responses I have for `get_nft_transfers`, there is `transaction_type` value `single`. I assume this is `multiple` for ERC1155 transactions involving more than one NFT?

laruh replied:
Yeah, I think the same. I asked moralis team about this, but thed didnt reply. 
![image](https://github.com/KomodoPlatform/komodo-docs-mdx/assets/52405288/8920233b-e46d-4086-98ae-d2cd30eecc3d)

Currently I use `safeTransferFrom` functions for both [erc721](https://eips.ethereum.org/EIPS/eip-721#specification) and [erc1155](https://eips.ethereum.org/EIPS/eip-1155#specification) standards. However, erc1155 supports `safeBatchTransferFrom` function, so  the second variant could be "Batch", but I dont know how moralis call it. well at least we use String for this field so we can catch it` pub(crate) transaction_type: String,`

smk762 commented:
Added the [refresh_nft_metadata and update_nft](https://github.com/KomodoPlatform/komodo-docs-mdx/pull/42/commits/5e86b5d9e57774c57d04520b22a78620d70b87b1) methods.
Updated the other methods to include filters and remove url param where applicable. 
Will "test the doc" tomorrow and generate some fresh / error responses

laruh commented on src/pages/atomicdex/api/v20-dev/non_fungible_tokens/index.mdx:
```diff
@@ -0,0 +1,762 @@
+
+- Get a list of your tokens with [get_nft_list](/atomicdex/api/v20-dev/non-fungible-tokens/get-nft-list)
+- Get a list of token transfers with [get_nft_transfers](/atomicdex/api/v20-dev/non-fungible-tokens/get-nft-transfers)
+- Get token metadata with [get_nft_metadata](/atomicdex/api/v20-dev/non-fungible-tokens/get-nft-metadata)
+
```
hmm I think we can add into this list `update_nft` and `refresh_nft_metadata` methods, bcz they are related to NFT information.

smk762 replied:
Links added in https://github.com/KomodoPlatform/komodo-docs-mdx/pull/42/commits/0e74d8b2e61dfa43e7db9136a05ece3b73ad6748
Full methods pending

laruh commented on src/pages/atomicdex/api/v20-dev/non_fungible_tokens/index.mdx:
```diff
@@ -0,0 +1,762 @@
+
+## Get a list of NFTs {{label : 'get_nft_list', tag : 'API-v2'}}
+
+Returns a list of the NFTs owned by the user, shown in descending order of the `block_number` value (the block height when the amount or owner changed). If the request is for NFTs on more than one chain, this means that the order may not be chronological.
```
`the block height when the amount or owner changed`

the note: if we are talking about amount, it can be changed if NFT has erc1155 standard type, but not always. 
when user receive more erc1155 tokens, then `block_number` rises, but when user transfer not all erc1155 tokens, usually `block_number` stays the same, bcz[ `safeTransferFrom` function](https://eips.ethereum.org/EIPS/eip-1155#specification) does not affect `block_number` changes. well potentially NFT creator can implement this functionality, but it increases fee cost. so usually from my experience `block_number` stayed the same after transferring some erc1155 amount.

smk762 replied:
updated in https://github.com/KomodoPlatform/komodo-docs-mdx/pull/42/commits/0e74d8b2e61dfa43e7db9136a05ece3b73ad6748

laruh commented on src/pages/atomicdex/api/v20-dev/non_fungible_tokens/index.mdx:
```diff
@@ -0,0 +1,762 @@
+| chains        | array            | List of chains to scan for NFTs.                                                                          |
+| max           | boolean          | Optional, defaults to `false`. If `true`, response will return all NFTs without pagination, and will ignore the `limit` and `page_number` values.               |
+| limit         | integer          | Optional, defaults to `10`. The number of NFTs displayed per page in response.                            |
+| page_number   | integer          | Optional, defaults to `0`. The page offset for items in response.                                         |
```
we use nonzero usize type for page_number, you will have this err, if you set 0 page

```
{"mmrpc":"2.0","error":"Error parsing request: invalid value: integer `0`, expected a nonzero usize","error_path":"dispatcher","error_trace":"dispatcher:109]","error_type":"InvalidRequest","error_data":"invalid value: integer `0`, expected a nonzero usize","id":null}
```

according to this implementation, it doesnt matter provided user 1 page or not. So we can say that 1st page is default.
```rust
let (offset, limit) = if max {
                (0, count_total)
            } else {
                match page_number {
                    Some(page) => ((page.get() - 1) * limit, limit),
                    None => (0, limit),
                }
            };
```

laruh reviewed (changes requested):
Thanks for the fixes! I have some notes, will provide more later.

laruh reviewed (changes requested):
Some additional notes

laruh commented:
[updated git](https://gist.github.com/laruh/248d652d88891f3cadb09fea3a441466)
contains changes from [this PR](https://github.com/KomodoPlatform/atomicDEX-API/pull/1877)

laruh commented on src/pages/atomicdex/api/v20-dev/non_fungible_tokens/index.mdx:
```diff
@@ -0,0 +1,762 @@
+| page_number   | integer          | Optional, defaults to `0`. The page offset for items in response.                                         |
+
+
+### Response Parameters
+
+| Parameter            | Type             | Description                                                                                      |
```
[nft list info](https://gist.github.com/laruh/248d652d88891f3cadb09fea3a441466#get_nft_list-request) 

Some updates:
In `struct Nft` and `struct NftCommon` you can see that a lot of fields have become optional. 
Please note that `NftCommon` will be flattened and in response in Nft object there will be shown all fields from `NftCommon` instead of `common` field.

Also `collection_name` field right now called just `name`, as well as `token_name` inside `uri_meta`. it is also called `name`.
That is, the external `name` is the name of the collection, the `name` field inside `UriMeta` is the token name.

In addition, notes about `"image"` and `"image_url"` from `"uri_meta"`.
Right now `image` is the original field from `"token_uri"` link.
`"image_url"` is the field where we collect images from `token_uri` and `metadata` fields. If `image` or `image_url` are null in `token_uri` link, then we try to get one of them from `metadata`. 

Another note:
`token_uri` and `metadata` usually contain either `image` or `image_url` with image url. But most often nft creators use only `image` name for this value (from my observation), less often they use both parameters with the same url.

laruh commented on src/pages/atomicdex/api/v20-dev/non_fungible_tokens/index.mdx:
```diff
@@ -0,0 +1,762 @@
+| filters       | object           | Optional. A standard [TokenFilter](/atomicdex/api/v20#token-filter) object.                               |
+
+
+### Response Parameters
+
+| Parameter            | Type             | Description                                                                                      |
```
[updated info about transfers](https://gist.github.com/laruh/248d652d88891f3cadb09fea3a441466#get_nft_transfers-request)

same as for nft object `struct NftTransferCommon` will be flattened in response.
Updates: some fields have became optional.
`token_uri` field was added in nft transfers response. its the same field as from Nft object. 
`collection_name` is the same field as external `name` field from Nft object.

laruh commented on src/pages/atomicdex/api/v20-dev/non_fungible_tokens/index.mdx:
```diff
@@ -0,0 +1,762 @@
+`verified` has no description. Related to https://cointelegraph.com/news/nft-whale-pranksy-pranked-by-fake-banksy-for-97-7-eth? Who verifies it? I can see there are ways to verify on opensea etc, I assume Moralis incormoprates this.
+What are the other possible values for `transaction_type`?
+What is `operator`?
+What does `verified` mean?
```
I finally got the answer about `verified`
Its for their devs as I understood, and actually they are going to remove it in the future.
![image](https://github.com/KomodoPlatform/komodo-docs-mdx/assets/52405288/53b78f20-028b-4219-892f-03b3120e2ba4)

laruh commented on src/pages/atomicdex/api/v20-dev/non_fungible_tokens/index.mdx:
```diff
@@ -0,0 +1,762 @@
+
+Returns a list of the NFTs owned by the user, shown in descending order of the `block_number` value (the block height when the amount or owner changed). If the request is for NFTs on more than one chain, this means that the order may not be chronological.
+
+### Request Parameters
+
```
has new param `protect_from_spam`
definition you can find in this [pr description](https://github.com/KomodoPlatform/atomicDEX-API/pull/1877) 

Also `get_nft_transfers` and `get_nft_metadata` requests have it.

If `protect_from_spam` param is `true` in `get_nft_list`, `get_nft_metadata`, `get_nft_transfers` reqs, then we will search urls in fields like `collection_name`, `token_name`, `symbol`. if they have it, then the text in these parameters will be replaced by `URL redacted for user protection`, and `possible_spam` in `Nft` or `NftTransferHistory` structures will be `true`

laruh reviewed (changes requested):
Notes related to NFT API from curent dev version (this nft api will be in [1.0.6 release](https://github.com/KomodoPlatform/atomicDEX-API/pull/1871))

smk762 commented on src/pages/atomicdex/api/v20/index.mdx:
```diff
@@ -431,6 +431,93 @@ For requests which return many results, pagination offsets may be applied. ** Us
+| token_name             | string  | Optional. Name of the token.                                                                                                      |
+| description            | string  | Optional. Description of the token.                                                                                               |
+| attributes             | object or array of objects  | Optional. The values within this parameter will vary, and are set by the creator. Often used to store traits. |
+| animation_url          | string  | Optional. Url to an animationto be displaye instead of a static image.                                                            |
```
fixed in https://github.com/KomodoPlatform/komodo-docs-mdx/commit/b55adc66393340f1292728366eb7eeada993c243

smk762 commented on src/pages/atomicdex/api/v20-dev/non_fungible_tokens/index.mdx:
```diff
@@ -0,0 +1,762 @@
+| block_number         | integer          | The block height when the amount or owner changed.                                               |
+| chain                | string           | Chain name. One of `AVALANCHE`, `BSC`, `ETH`, `FANTOM`, or `POLYGON`.                            |
+| collection_name      | string           | May be `null`. An NFT collection name.                                                                          |
+| contract_type        | string           | The type of NFT contract standard. One of `ERC721` or `ERC1155`.                                 |
```
Is there a case where a user would be able to transfer these nfts into our wallet in the first place?

smk762 commented on src/pages/atomicdex/api/v20-dev/non_fungible_tokens/index.mdx:
```diff
@@ -0,0 +1,762 @@
+| last_token_uri_sync  | string           | When the token_uri was last updated.                                                             |
+| last_metadata_sync   | string           | When the metadata was last updated.                                                              |
+| metadata             | string           | The metadata of the token.                                                                       |
+| minter_address       | string           | Minter address.                                                                                  |
```
minter address?

smk762 commented on src/pages/atomicdex/api/v20-dev/non_fungible_tokens/index.mdx:
```diff
@@ -0,0 +1,762 @@
+| chains        | array            | List of chains to scan for NFTs.                                                                          |
+| max           | boolean          | Optional, defaults to `false`. If `true`, response will return all NFTs without pagination, and will ignore the `limit` and `page_number` values.               |
+| limit         | integer          | Optional, defaults to `10`. The number of NFTs displayed per page in response.                            |
+| page_number   | integer          | Optional, defaults to `0`. The page offset for items in response.                                         |
```
fixed in https://github.com/KomodoPlatform/komodo-docs-mdx/commit/b55adc66393340f1292728366eb7eeada993c243

smk762 commented on src/pages/atomicdex/api/v20-dev/non_fungible_tokens/index.mdx:
```diff
@@ -0,0 +1,762 @@
+| Parameter     | Type             | Description                                                                                               |
+| ------------- | ---------------- | ----------------------------------------------------------------------------------------------------------|
+| chains        | array            | List of chains to scan for NFTs.                                                                          |
+| max           | boolean          | Optional, defaults to `false`. If `true`, response will return all NFTs without pagination, and will ignore the `limit` and `page_number` values.               |
```
Fixed https://github.com/KomodoPlatform/komodo-docs-mdx/pull/42/commits/0ee610bc3ef7f241fc19818600d8bfbf98967ff2

smk762 commented on src/pages/atomicdex/api/v20-dev/non_fungible_tokens/index.mdx:
```diff
@@ -0,0 +1,762 @@
+| last_token_uri_sync  | string           | Last time the token URI was synced.                                                              |
+| last_metadata_sync   | string           | Last time the token metadata was synced.                                                         |
+| minter_address       | string           | Minter address.                                                                                  |
+| possible_spam        | boolean          | Possible spam.                                                                                   |
```
all or just `possible_spam`?

smk762 commented on src/pages/atomicdex/api/v20/index.mdx:
```diff
@@ -431,6 +431,93 @@ For requests which return many results, pagination offsets may be applied. ** Us
+| receive                | boolean | Optional, defaults to `false`. If `true`, only transfers where user received NFTs are included in the response.                     |
+| send                   | boolean | Optional, defaults to `false`. If `true`, only transfers where user sent NFTs are included in the response.                         |
+| from_date              | string  | Optional. A UTC date string in the format `YYYY-MM-DD HH:MM:SS.SSS`. Only transfers after this date are included in the response.   |
+| to_date                | string  | Optional. A UTC date string in the format `YYYY-MM-DD HH:MM:SS.SSS`. Only transfers before this date are included in the response.  |
```
has been fixed

laruh commented on src/pages/atomicdex/api/v20/index.mdx:
```diff
@@ -554,3 +658,117 @@ The `TokensRequest` object includes the following items for a given coin or toke
+| withdraw\_data | object | A standard WithdrawDataErc721 or WithdrawDataErc1155 object.        |
+
+<DevComment>
+  TODO:
+  sections pointed by the slugs don't exist: `[WithdrawDataErc721](/atomicdex/api/v20#withdraw-data-erc721) or [WithdrawDataErc1155](/atomicdex/api/v20#withdraw-data-erc155)`
+</DevComment>
```
this todo left

smk762 commented:
@kivqa please review the docs by testing each example to confirm it is working as expected. Feel free to add additional examples using optional params or to try generating extra error responses to add to the bottom of the method docs.

laruh commented on src/pages/atomicdex/api/v20-dev/non_fungible_tokens/get_nft_list/index.mdx:
```diff
@@ -0,0 +1,322 @@
+          }
+        }
+      ],
+      "skipped": 1,
```
"skipped": 0, as spam BSC token was not excluded in result list of items.

laruh commented on src/pages/atomicdex/api/v20-dev/non_fungible_tokens/get_nft_list/index.mdx:
```diff
@@ -0,0 +1,322 @@
+| protect\_from\_spam | boolean | Optional, defaults to `false`. If `true`, any potentially spam link found in collection name, token name, symbol will be censored                 |
+| filters             | object  | Optional. A standard [TokenFilter](/atomicdex/api/v20/#token-filter) object.                                                                      |
+
+### Response Parameters
```
we should include in response param description `skipped` and `total` fields, as actual response is this object
```rust
pub struct NftList {
    pub(crate) nfts: Vec<Nft>,
    pub(crate) skipped: usize,
    pub(crate) total: usize,
}
```

smk762 replied:
thanks. Added, and the nested object moved to common structures doc

laruh reviewed (changes requested):
Thanks for moving RPC to separate folders, now its more convenient!
Added first notes for `get_nft_list`

laruh commented on src/pages/atomicdex/api/v20-dev/non_fungible_tokens/get_nft_transfers/index.mdx:
```diff
@@ -0,0 +1,125 @@
+| to\_address        | string  | Address of new owner which received the token(s).                                                                                                                     |
+| status             | string  | Transfer status. Will be either `Send` or `Receive`. When the `from_address` and `to_address` are the same (i.e. sending to yourself), this value will be `Receive`.  |
+| verified           | integer | May be `null`. A deprecated field which will be removed in future.                                                                                                    |
+| operator           | string  | May be `null`.                                                                                                                                                        |
```
need to add two more fields

```rust
pub(crate) fee_details: Option<EthTxFeeDetails>,
pub(crate) confirmations: u64,
 ```
 
 ```rust
 pub struct EthTxFeeDetails {
    pub coin: String,
    pub gas: u64,
    /// WEI units per 1 gas
    pub gas_price: BigDecimal,
    pub total_fee: BigDecimal,
}
```
gas - the total amount of gas used for a transaction
gas_price - gas price
total_fee - the total cost of the transaction

UPD: `fee_details` in get transfers is actually `WithdrawFee` object.
But it is optional, as we can get None in eth api response for old transfers ([see PR description](https://github.com/KomodoPlatform/komodo-defi-framework/pull/1989))

smk762 replied:
Added

laruh replied:
Oops, sorry, missed the part that in `get_nft_transfers` there is no `type` field, as we have in `WithdrawFee`, so fee info for `get_nft_transfers` will be just that

```
| Parameter       | Type           | Description                                                                       |
| --------------- | -------------- | --------------------------------------------------------------------------------- |
| coin            | string         | The coin which will be used to pay the transaction fee.                           |
| gas             | integer        | The amount of gas to be used for the transaction.                                 |
| gas\_price      | numeric string | Price per unit of gas to be used for the transaction.                             |
| total\_fee      | numeric string | Gas price multiplied by gas amount.                                               |
```
This object can be located below `NftTransfer` object

laruh replied:
Oops, sorry, missed the part that in `get_nft_transfers` there is no `type` field, as we have in `WithdrawFee`, so fee info for `get_nft_transfers` will be just that

```
| Parameter       | Type           | Description                                                                       |
| --------------- | -------------- | --------------------------------------------------------------------------------- |
| coin            | string         | The coin which will be used to pay the transaction fee.                           |
| gas             | integer        | The amount of gas to be used for the transaction.                                 |
| gas\_price      | numeric string | Price per unit of gas to be used for the transaction.                             |
| total\_fee      | numeric string | Gas price multiplied by gas amount.                                               |
```
This object can be located in `dex/api/v20/index.mdx`with other nft objects like `TokenMetadata`, `TokenTransferFilter` etc

laruh commented on src/pages/atomicdex/api/v20-dev/non_fungible_tokens/refresh_nft_metadata/index.mdx:
```diff
@@ -0,0 +1,58 @@
+
+| Parameter      | Type   | Description                                                                                                                                             |
+| -------------- | ------ | ------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| chains         | string | Chains which holds the NFT you would like to updated metadata for.                                                                                      |
```
just `chain`

smk762 replied:
tanks

laruh commented on src/pages/atomicdex/api/v20-dev/non_fungible_tokens/withdraw_nft/index.mdx:
```diff
@@ -0,0 +1,298 @@
+    "method": "withdraw_nft",
+    "mmrpc": "2.0",
+    "params": {
+      "type": "withdraw_erc721",
```
You repeated ERC721 withdraw in ERC1155 Withdraw Example

laruh commented on src/pages/atomicdex/api/v20-dev/non_fungible_tokens/withdraw_nft/index.mdx:
````diff
@@ -0,0 +1,298 @@
+<CodeGroup title="Withdraw NFTs" tag="POST" label="withdraw_nft" mm2MethodDecorate="true">
+  ```json
+  {
+    "userpass": "testpsw",
+    "method": "withdraw_nft",
+    "mmrpc": "2.0",
+    "params": {
+      "type": "withdraw_erc1155",
+      "withdraw_data": {
+        "chain": "POLYGON",
+        "to": "0x27Ad1F808c1ef82626277Ae38998AfA539565660",
+        "token_address": "0x2953399124f0cbb46d2cbacd8a89cf0599974963",
+        "token_id": "110473361632261669912565539602449606788298723469812631769659886404530570536720",
+        "amount": "1"
````
No `max: true` in ERC1155 Withdraw Max Example

laruh commented on src/pages/atomicdex/api/v20/index.mdx:
```diff
@@ -101,3 +101,647 @@ It includes a uniform request, successful and error response formats. At the mom
+| token\_address | string  | Token address.                                                                                                                           |
+| token\_id      | string  | Token ID.                                                                                                                                |
+| fee            | string  | Optional. A standard [WithdrawFee](/atomicdex/api/v20/#withdraw-fee) object.                                                             |
+| amount         | string  | Optional, ERC1155 only. Defaults to `1`. Amount of NFTs to withdraw. Ignored if `max` is \`true.                                         |
```
Should `true` be written like this \ `true in docs ?

gcharang replied:
that's a typo. should be `true` cc: @smk762

smk762 replied:
fixed

laruh reviewed (changes requested):
some more notes

smk762 commented on src/pages/atomicdex/api/v20-dev/non_fungible_tokens/index.mdx:
```diff
@@ -0,0 +1,762 @@
+| amount               | string           | Amount of tokens transferred.                                                                    |
+| block_hash           | string           | Hash of block in which transfer occurred.                                                        |
+| block_number         | integer          | Height of block in which transfer occurred.                                                      |
+| block_timestamp      | string           | Block date/time string.                                                                          |
```
updated

smk762 commented on src/pages/atomicdex/api/v20-dev/non_fungible_tokens/get_nft_list/index.mdx:
```diff
@@ -0,0 +1,322 @@
+| max                 | boolean | Optional, defaults to `false`. If `true`, response will return all NFTs without pagination, and will ignore the `limit` and `page_number` values. |
+| limit               | integer | Optional, defaults to `10`. The number of NFTs displayed per page in response.                                                                    |
+| page\_number        | integer | Optional, defaults to `1`. The page offset for items in response.                                                                                 |
+| protect\_from\_spam | boolean | Optional, defaults to `false`. If `true`, any potentially spam link found in collection name, token name, symbol will be censored                 |
```
done

smk762 commented on src/pages/atomicdex/api/v20-dev/non_fungible_tokens/get_nft_metadata/index.mdx:
```diff
@@ -0,0 +1,115 @@
+| chain          | string | The token chain.   |
+| token\_address | string | The token address. |
+| token\_id      | string | Token ID.          |
+
```
added

laruh replied:
Please also add `URL redacted for user protection` info as in `get_nft_list` request

smk762 commented on src/pages/atomicdex/api/v20-dev/non_fungible_tokens/get_nft_transfers/index.mdx:
```diff
@@ -0,0 +1,125 @@
+| protect\_from\_spam | boolean | Optional, defaults to `false`. If `true`, any potentially spam link found in `collection_name`, `token_name` will be censored                              |
+| filters             | object  | Optional. A standard [TokenTransferFilter](/atomicdex/api/v20/#token-transfer-filter) object.                                                              |
+
+### Response Parameters
```
thanks. fields added and nested structure moved to commons

smk762 commented on src/pages/atomicdex/api/v20-dev/non_fungible_tokens/get_nft_transfers/index.mdx:
```diff
@@ -0,0 +1,125 @@
+  What does `verified` mean?
+</DevComment>
+
+#### 📌 Example with date and `send` filters
```
copy and paste applied

smk762 commented:
Thanks for the feedback, all items addressed. I've added a new section for the NFT databases, please check the descriptions are accurate.

laruh commented on src/pages/atomicdex/api/v20-dev/non_fungible_tokens/get_nft_list/index.mdx:
```diff
@@ -0,0 +1,323 @@
+| Parameter | Type            | Description                                                                            |
+| --------- | --------------- | -------------------------------------------------------------------------------------- |
+| nfts      | list of objects | A list of standard [NftInfo](/atomicdex/api/common_structures/nfts/#nft-info) objects. |
+| total     | string          | The total number of NFTs in your wallet matching the request filters.                  |
+| skipped   | string          | The number of NFTs in your wallet excluded by the request filters.                     |
```
skipped and total are integers

smk762 replied:
fixed

laruh replied:
![image](https://github.com/KomodoPlatform/komodo-docs-mdx/assets/52405288/0f218223-6b1d-4daa-9e99-d3b5de5a3e6b)
I see that there are two files where we have strings. Are these files used in docs?

laruh commented on src/pages/atomicdex/api/v20/index.mdx:
```diff
@@ -101,3 +101,647 @@ It includes a uniform request, successful and error response formats. At the mom
+| to             | string  | Destination address to withdraw the token to.                                                                                            |
+| token\_address | string  | Token address.                                                                                                                           |
+| token\_id      | string  | Token ID.                                                                                                                                |
+| fee            | object  | A standard [WithdrawFee](/atomicdex/api/v20/#withdraw-fee) object. May be missing for older transfers.                                   |
```
Seems like `[WithdrawFee](/atomicdex/api/v20/#withdraw-fee)` is an object, which we have in withdraw fungible token **response**. 

In nft withdraw request I used `fee: Option<WithdrawFee>`(rust version), which is used in coins withdraw **requests**.
```rust
#[serde(tag = "type")]
pub enum WithdrawFee {
    UtxoFixed {
        amount: BigDecimal,
    },
    UtxoPerKbyte {
        amount: BigDecimal,
    },
    EthGas {
        /// in gwei
        gas_price: BigDecimal,
        gas: u64,
    },
    Qrc20Gas {
        /// in satoshi
        gas_limit: u64,
        gas_price: u64,
    },
    CosmosGas {
        gas_limit: u64,
        gas_price: f64,
    },
}
```
Looks like this [object](https://github.com/KomodoPlatform/komodo-docs-mdx/blob/add-nft-docs/src/pages/atomicdex/api/common_structures/index.mdx#L150) . Actually there is a repetition in another [place](https://github.com/KomodoPlatform/komodo-docs-mdx/blob/add-nft-docs/src/pages/atomicdex/api/v20/index.mdx#L381). I think we should leave one example in `common_structures` folder.


`The FeeInfo response object includes the following items for [withdraw (v2)](/atomicdex/api/v20/withdraw/) requests`

`WithdrawFee` enum, which is `FeeInfo` in docs, is for requests, not response
```rust
pub struct WithdrawRequest {
    coin: String,
    from: Option<WithdrawFrom>,
    to: String,
    #[serde(default)]
    amount: BigDecimal,
    #[serde(default)]
    max: bool,
    fee: Option<WithdrawFee>,
    memo: Option<String>,
    /// Currently, this flag is used by ETH/ERC20 coins activated with MetaMask **only**.
    #[cfg(target_arch = "wasm32")]
    #[serde(default)]
    broadcast: bool,
}
```


and this `TxFeeDetails` enum we have in responses, it is `WithdrawFee` in your docs.
```rust
/// Please note that no type should have the same structure as another type,
/// because this enum has the `untagged` deserialization.
#[derive(Clone, Debug, PartialEq, Serialize)]
#[serde(tag = "type")]
pub enum TxFeeDetails {
    Utxo(UtxoFeeDetails),
    Eth(EthTxFeeDetails),
    Qrc20(Qrc20FeeDetails),
    Slp(SlpFeeDetails),
    Tendermint(TendermintFeeDetails),
    #[cfg(all(
        feature = "enable-solana",
        not(target_os = "ios"),
        not(target_os = "android"),
        not(target_arch = "wasm32")
    ))]
    Solana(SolanaFeeDetails),
}
```

smk762 replied:
updated. thanks, there were a half a dozen or so other duplicated structures. I've removed all that was in the `dev` and `v2-dev` folders and added them where appropriate under the `common_structures` folder. The `FeeInfo` object used for responses and the `WithdrawFee` object used for requests shoudl now be properly applied within the NFT docs.

laruh replied:
@smk762 

`FeeInfo` should be for requests, 
https://github.com/KomodoPlatform/komodo-docs-mdx/blob/418d2d443d11271605bab424d4a6cc30313bfe89/data-for-gpts/komodefi-api/all-api-content.txt#L341

Here in `WithdrawNftData` in docs we have `WithdrawFee`, but in rust repo `WithdrawNftData` is for reqs. `FeeInfo` should be her.
https://github.com/KomodoPlatform/komodo-docs-mdx/blob/418d2d443d11271605bab424d4a6cc30313bfe89/data-for-gpts/komodefi-api/v20-api-content.txt#L3321-L3331

and `WithdrawFee` should be for response.
https://github.com/KomodoPlatform/komodo-docs-mdx/blob/418d2d443d11271605bab424d4a6cc30313bfe89/data-for-gpts/komodefi-api/v20-api-content.txt#L3256-L3270
Thus `| gas             | integer        | `Eth` type only. The amount of gas to be used for the transaction.` not to be used, but gas was used

Just in rust code we have another names.
rust `WithdrawFee` enum is for req
```rust
#[derive(Clone, Debug, Deserialize, PartialEq)]
#[serde(tag = "type")]
pub enum WithdrawFee {
    UtxoFixed {
        amount: BigDecimal,
    },
    UtxoPerKbyte {
        amount: BigDecimal,
    },
    EthGas {
        /// in gwei
        gas_price: BigDecimal,
        gas: u64,
    },
    Qrc20Gas {
        /// in satoshi
        gas_limit: u64,
        gas_price: u64,
    },
    CosmosGas {
        gas_limit: u64,
        gas_price: f64,
    },
}
```

This is for responses
```rust
#[derive(Clone, Debug, PartialEq, Serialize)]
#[serde(tag = "type")]
pub enum TxFeeDetails {
    Utxo(UtxoFeeDetails),
    Eth(EthTxFeeDetails),
    Qrc20(Qrc20FeeDetails),
    Slp(SlpFeeDetails),
    Tendermint(TendermintFeeDetails),
    #[cfg(all(
        feature = "enable-solana",
        not(target_os = "ios"),
        not(target_os = "android"),
        not(target_arch = "wasm32")
    ))]
    Solana(SolanaFeeDetails),
}


// for example eth tx fee
pub struct EthTxFeeDetails {
    pub coin: String,
    pub gas: u64,
    /// WEI units per 1 gas
    pub gas_price: BigDecimal,
    pub total_fee: BigDecimal,
}
```

I think you should rename `WithdrawFee` type in docs to `TxFeeDetails`, and `FeeInfo` rename to `WithdrawFee`. So it will be easier for all of us, as we will have same object names in both rust code and docsx. Se you and me can check what fee object should be in docs according rust code name.

laruh reviewed (changes requested):
Some new notes

laruh commented on filepathSlugs.json:
```diff
@@ -2788,6 +2829,67 @@
+    "nft-list-tables",
+    "example-coin-nft-list-table-query",
+    "nft-transfer-table",
+    "example-scanned-nft-blocks-table-query",
```
should be `"example-nft-transfer-table-query"` or smth like this, but not scanned-nft-blocks, as above is nft transfer table.


Also there is some problem with sql command display
![image](https://github.com/KomodoPlatform/komodo-docs-mdx/assets/52405288/ed7efa77-bb08-4822-b580-06b26840104e)

I mean example coin_nft_list table query looks differently
![image](https://github.com/KomodoPlatform/komodo-docs-mdx/assets/52405288/3ded9cf8-a2d0-4d07-9ddc-6bff5e5aabb8)

smk762 replied:
these slugs are automatically generated for internal anchor links, so I wont mess with it.

gcharang replied:
fixed sql command formats in [`3cd7280` (#42)](https://github.com/KomodoPlatform/komodo-docs-mdx/pull/42/commits/3cd72803054e40e352df7d7f452816e539842e10)

gcharang replied:
@laruh @smk762 can you please explain what the issue with the autogenerated slugs is?

gcharang replied:
I get it, the headers are repeating. Need to change one of them. fixed in [`e0267a3` (#42)](https://github.com/KomodoPlatform/komodo-docs-mdx/pull/42/commits/e0267a30638996e8260d3c5b5032285619c4562e)

smk762 commented on src/pages/atomicdex/api/v20-dev/non_fungible_tokens/get_nft_transfers/index.mdx:
```diff
@@ -0,0 +1,103 @@
+| Parameter         | Type            | Description                                                                                    |
+| ----------------- | --------------- | ---------------------------------------------------------------------------------------------- |
+| transfer\_history | list of objects | A list of standard [NftTransfer](/atomicdex/api/common_structures/nfts/#nft-transfer) objects. |
+| total             | string          | The total number of NFT transfers in your wallet matching the request filters.                 |
+| skipped           | string          | The number of NFT transfers in your wallet excluded by the request filters.                    |
```
fixed

laruh replied:
![image](https://github.com/KomodoPlatform/komodo-docs-mdx/assets/52405288/6918c57c-c804-4f66-9536-ea4c8957859c)
Same here, do we need another files, where there are still string values?

smk762 commented on src/pages/atomicdex/api/v20-dev/non_fungible_tokens/get_nft_transfers/index.mdx:
```diff
@@ -0,0 +1,103 @@
+| max                 | boolean | Optional, defaults to `false`. If `true`, response will return all NFT transfers without pagination, and will ignore the `limit` and `page_number` values. |
+| limit               | integer | Optional, defaults to `10`. The number of NFT transfers displayed per page in response.                                                                    |
+| page\_number        | integer | Optional, defaults to `1`. The page offset for items in response.                                                                                          |
+| protect\_from\_spam | boolean | Optional, defaults to `false`. If `true`, any potentially spam link found in `collection_name`, `token_name` will be censored                              |
```
done

smk762 commented on src/pages/atomicdex/api/common_structures/nfts/index.mdx:
```diff
@@ -20,3 +20,71 @@ The `TokensRequest` object includes the following items for a given coin or toke
+| verified           | integer | May be `null`. A deprecated field which will be removed in future.                                                                                                    |
+| operator           | string  | May be `null`.                                                                                                                                                        |
+| fee\_details       | object  | Optional. A standard [WithdrawFee](/atomicdex/api/v20/#withdraw-fee) object.                                                                                          |
+
```
added

smk762 commented:
@gcharang @kivqa Ready for Review
@laruh These docs may need some further work, but we will merge it in soon so we can apply the rebranding from atomicdex in a final PR before publishing without leaving this PR subject to numerous conflicts.
Once rebranding PR is done, review and polish of this PR (alongside the HD Wallet PR) will be top of the list for docs tasks.

smk762 commented:
Note: The recently added `clear_nft_db` method specifies the `chains` parameter as optional. Currently it is required, but will be made optional later based on testing feedback provided to @laruh
